docker/docker.go

```diff
@@ -217,30 +217,52 @@ func DoRemoveContainer(ctx context.Context, dClient *client.Client, containerNam
 	return nil
 }
 
-func IsContainerRunning(ctx context.Context, dClient *client.Client, hostname string, containerName string, all bool) (bool, error) {
+func FindContainer(ctx context.Context, dClient *client.Client, hostname string, containerName string, all bool) (*types.Container, error) {
 	if dClient == nil {
-		return false, fmt.Errorf("Failed to check if container is running: docker client is nil for container [%s] on host [%s]", containerName, hostname)
+		return nil, fmt.Errorf("Failed to find container: docker client is nil for container [%s] on host [%s]", containerName, hostname)
 	}
 	var containers []types.Container
 	var err error
 	for i := 1; i <= RetryCount; i++ {
-		logrus.Infof("Checking if container [%s] is running on host [%s], try #%d", containerName, hostname, i)
+		logrus.Infof("Finding container [%s] on host [%s], try #%d", containerName, hostname, i)
 		containers, err = dClient.ContainerList(ctx, types.ContainerListOptions{All: all})
 		if err != nil {
-			logrus.Warnf("Error checking if container [%s] is running on host [%s]: %v", containerName, hostname, err)
+			logrus.Warnf("Error finding container [%s] exists on host [%s]: %v", containerName, hostname, err)
 			continue
 		}
 		break
 	}
 	if err != nil {
-		return false, fmt.Errorf("Error checking if container [%s] is running on host [%s]: %v", containerName, hostname, err)
+		return nil, fmt.Errorf("Error checking if container [%s] exists on host [%s]: %v", containerName, hostname, err)
 	}
 	for _, container := range containers {
 		if len(container.Names) != 0 && container.Names[0] == "/"+containerName {
-			return true, nil
+			return &container, nil
 		}
 	}
-	return false, nil
+	return nil, nil
+}
+
+func DoesContainerExist(ctx context.Context, dClient *client.Client, hostname string, containerName string, all bool) (bool, error) {
+	if dClient == nil {
+		return false, fmt.Errorf("Failed to check if container exists: docker client is nil for container [%s] on host [%s]", containerName, hostname)
+	}
+	container, err := FindContainer(ctx, dClient, hostname, containerName, all)
+	if err != nil {
+		return false, fmt.Errorf("Error checking if container [%s] is running on host [%s]: %v", containerName, hostname, err)
+	}
+	return container != nil, nil
+}
+
+func IsContainerRunning(ctx context.Context, dClient *client.Client, hostname string, containerName string, all bool) (bool, error) {
+	if dClient == nil {
+		return false, fmt.Errorf("Failed to check if container is running: docker client is nil for container [%s] on host [%s]", containerName, hostname)
+	}
+	container, err := FindContainer(ctx, dClient, hostname, containerName, all)
+	if err != nil {
+		return false, fmt.Errorf("Error checking if container [%s] is running on host [%s]: %v", containerName, hostname, err)
+	}
+	return container.State == "running", nil
```

Comment:
This could cause a runtime error in the current implementation because the `FindContainer` function could return "nil, nil"

Reply:
Good catch